app/controllers/case_information_controller.rb

```diff
@@ -97,10 +110,121 @@ def nomis_offender_id_from_url
   end
 
+  def stage1_errors
+    if case_information_params[:last_known_location].nil?
+      @case_info.errors.messages.reject! do |f, _m|
+        f != :probation_service && f != :last_known_location
+      end
+      display_address_page
+    elsif case_information_params[:last_known_location] == 'Yes' && case_information_params[:probation_service].nil?
+      @case_info.errors.messages.select! do |f, _m|
+        f == :probation_service
+      end
+      display_address_page
+    end
+  end
+
+  def display_address_page
+    @prisoner = prisoner(case_information_params[:nomis_offender_id])
+    render :new
+  end
+
+  def handle_stage1
+    if @case_info.scottish_or_ni?
+      @case_info.save_scottish_or_ni
+    elsif @case_info.english_or_welsh?
+      @case_info.probation_service = 'England' if @case_info.last_known_location == 'No'
+      @prisoner = prisoner(case_information_params[:nomis_offender_id])
+      @case_info.errors.clear
+      render :missing_info
+    else
+      stage1_errors
+    end
+  end
+
+  def handle_stage2
+    @case_info.probation_service = 'England' if @case_info.english?
+
+    unless @case_info.stage2_filled?
+      @prisoner = prisoner(case_information_params[:nomis_offender_id])
+      @case_info.errors.messages.reject! do |f, _m|
+        [:welsh_offender, :probation_service].include?(f)
+      end
+      render :missing_info
+    end
+  end
+
+  def case_information_updated?
+    if ['Scotland', 'Northern Ireland'].include?(case_information_params[:probation_service]) &&
+      case_information_params[:last_known_location] == 'Yes'
+      @case_info.update(probation_service: case_information_params[:probation_service], tier: 'N/A',
+                        case_allocation: 'N/A', team: nil, welsh_offender: 'No', manual_entry: true)
+    else
+      probation = if case_information_params[:last_known_location] == 'No'
+                    'England'
+                  else
+                    case_information_params[:probation_service]
+                  end
+
+      welsh = welsh_offender?(probation: probation)
+      @case_info.update(probation_service: probation, tier: case_information_params[:tier],
+                        case_allocation: case_information_params[:case_allocation],
+                        team_id: team_identifier,  welsh_offender: welsh, manual_entry: true)
+    end
+  end
+
+  def team_identifier
+    Team.find_by(name: params['input-autocomplete'])&.id
+  end
+
+  def welsh_offender?(probation:)
+    if probation == 'Wales'
+      @case_info.welsh_offender = 'Yes'
+    else
+      @case_info.welsh_offender = 'No'
+    end
+  end
+
+  def send_email
+    return unless @case_info.probation_service == 'England' || @case_info.probation_service == 'Wales'
+
+    prepare_email
+  end
+
+  def prepare_email
+    spo = Nomis::Elite2::UserApi.user_details(current_user).email_address.try(:first)
+    ldu = @case_info.team.try(:local_divisional_unit)
+    emails = [ldu.try(:email_address), spo]
```

Comment:
would be more idiomatic to use 'compact' or reject(&:blank?) to avoid the next on line 203